### PR TITLE
git: improve sparse checkout support

### PR DESCRIPTION
**Pull request by bk2204**
When we invoke `git ls-files` to try to find all LFS files, we don't honour sparse file paths or exclusions.  While we should never actually traverse excluded files, using the `--exclude-standard` option can avoid loading some data with filtered clones, which may result in less data being downloaded.

In addition, we can honour sparse checkouts, since this code path is only used to handle the working tree and we know that the only files we need to consider are those Git actually put in the working tree.  The `--sparse` option is new in 2.35, but we already require 2.42 above, so we can use it unconditionally.

### CODE REVIEW

**Comment by bk2204** [comment]
My apologies, this slipped off my radar.  I've rebased this PR to make CI run again, and assuming it goes green, I'll mark it for review.

**Review by chrisd8088 — COMMENTED** [review]
This looks good, thank you!  I'd love to squeeze this into the upcoming v3.6.0 release.

My only reservation is that we don't have many tests using sparse checkouts (just a [pair](https://github.com/git-lfs/git-lfs/blob/3990c7ab4fa6c99540af3f73f33da47a72df08af/t/t-multiple-remotes.sh#L102-L113) of [tests](https://github.com/git-lfs/git-lfs/blob/3990c7ab4fa6c99540af3f73f33da47a72df08af/t/t-multiple-remotes.sh#L170-L181) in one script), and none which exercise this change in particular.

Of course we have lots of tests of `git lfs checkout` and `git lfs pull`, the only callers of the `ScanLFSFiles()` function, which is the only code path by which the `LsFilesLFS()` function is used, so we know this change it doesn't break anything, which is the most important point to confirm.

But it would also be nice to validate that it has the desired effect of increasing the efficiency of the `git ls-files` command when a user has a sparse checkout with sparse index.

I managed to write one test, which I could push to this PR's branch, or if you'd prefer to just include it in a commit, that's cool too.  I figured it might go into the `t/t-pull.sh` test script.  I've confirmed that it fails without the changes from this PR.

I'll try to craft a similar version for the `t/t-checkout.sh` script which exercises the `git lfs checkout` command instead of `git lfs pull`, unless you beat me to it!

```
begin_test "pull with partial clone and sparse checkout and index"
(
  set -e

  # Only test with Git version 2.42.0 as it introduced support for the
  # "objecttype" format option to the "git ls-files" command, which our
  # code requires.
  ensure_git_version_isnt "$VERSION_LOWER" "2.42.0"

  reponame="$(basename "$0" ".sh")"
  setup_remote_repo "$reponame"

  clone_repo "$reponame" repo

  git lfs track "*.dat"

  contents1="a"
  contents1_oid=$(calc_oid "$contents1")
  contents2="b"
  contents2_oid=$(calc_oid "$contents2")
  contents3="c"
  contents3_oid=$(calc_oid "$contents3")

  mkdir in out
  printf "%s" "$contents1" > a.dat
  printf "%s" "$contents2" > in/b.dat
  printf "%s" "$contents3" > out/c.dat
  git add .
  git commit -m "add files"

  git push origin main

  assert_server_object "$reponame" "$contents1_oid"
  assert_server_object "$reponame" "$contents2_oid"
  assert_server_object "$reponame" "$contents3_oid"

  # Create a partial clone with a cone-mode sparse checkout of one directory
  # and a sparse index, which is important because otherwise the "git ls-files" 
  # command ignores the --sparse option and lists all LFS files.
  cd ..
  git clone --filter=tree:0 --depth=1 --no-checkout \
    "$GITSERVER/$reponame" partial

  cd partial
  git sparse-checkout init --cone --sparse-index
  git sparse-checkout set in
  git checkout main

  assert_local_object "$contents1_oid" 1
  assert_local_object "$contents2_oid" 1
  refute_local_object "$contents3_oid"

  git lfs pull 2>&1 | tee pull.log
  grep -q "Downloading LFS objects" pull.log && exit 1

  refute_local_object "$contents3_oid"
)
end_test
```

File: git/git.go
Comment (bk2204) [reply git/git.go]: Done.

**Comment by bk2204** [comment]
I've squashed that test into the PR with a co-author credit and added a test for `git lfs checkout` as well (which also fails without these changes).

**Review by chrisd8088 — APPROVED** [review]
Thanks very much for collaborating on some tests for this PR!

I'm going to take the slightly unorthodox step of approving this PR without a second review from another @git-lfs/core team member, in the interest of trying to get to a v3.6.0 release PR quite soon after this is merged.

I think that's OK because (a) I only contributed to the additional tests in this PR, not the change to the client's code, (b) the tests pass CI, (c) we've both looked at the tests and made revisions, and (d) I've checked that each of the relevant assertions in the tests fails as expected if the client code change is removed, not just the first such assertion in each test.

**Comment by manoraj** [comment]
> in the interest of trying to get to a v3.6.0 release PR quite soon after this is merged

Hi @chrisd8088 May I know when v3.6.0 is gonna released?

**Comment by chrisd8088** [comment]
> May I know when v3.6.0 is gonna released?

@manoraj — You're in luck!  As it so happens, we're planning to release v3.6.0 later this week, assuming there are no unexpected issues.  See PR #5916 for details.

**Pull request merged**
